golem-rust/src/transaction.rs

```diff
@@ -0,0 +1,577 @@
+        }
+    }
+}
+
+impl<In: Clone, Out, Err> Operation for FnOperation<In, Out, Err> {
+    type In = In;
+    type Out = Out;
+    type Err = Err;
+
+    fn execute(&self, input: In) -> Result<Out, Err> {
+        (self.execute_fn)(input)
+    }
+
+    fn compensate(&self, input: In) -> Result<(), Err> {
+        (self.compensate_fn)(input)
+    }
+}
+
+#[allow(clippy::type_complexity)]
+struct FnOperationWithResult<In, Out, Err> {
+    execute_fn: Rc<dyn Fn(In) -> Result<Out, Err>>,
+    compensate_fn: Rc<dyn Fn(In, Result<Out, Err>) -> Result<(), Err>>,
+}
+
+impl<In, Out, Err> Clone for FnOperationWithResult<In, Out, Err> {
+    fn clone(&self) -> Self {
+        Self {
+            execute_fn: self.execute_fn.clone(),
+            compensate_fn: self.compensate_fn.clone(),
+        }
+    }
+}
+
+impl<In: Clone, Out, Err> Operation for FnOperationWithResult<In, Out, Err> {
+    type In = In;
+    type Out = Out;
+    type Err = Err;
+
+    fn execute(&self, input: In) -> Result<Out, Err> {
+        (self.execute_fn)(input)
+    }
+
+    fn compensate(&self, _input: Self::In) -> Result<(), Self::Err> {
+        Ok(())
+    }
+
+    fn compensate_with_result(&self, input: In, result: Result<Out, Err>) -> Result<(), Err> {
+        (self.compensate_fn)(input, result)
+    }
+}
+
+/// The result of a transaction execution.
+pub type TransactionResult<Out, Err> = Result<Out, TransactionFailure<Err>>;
+
+/// The result of a transaction execution that failed.
+#[derive(Debug)]
+pub enum TransactionFailure<Err> {
+    /// One of the operations failed with an error, and the transaction was fully rolled back.
+    FailedAndRolledBackCompletely(Err),
+    /// One of the operations failed with an error, and the transaction was partially rolled back
+    /// because the compensation action of one of the operations also failed.
+    FailedAndRolledBackPartially {
+        failure: Err,
+        compensation_failure: Err,
+    },
+}
+
+impl<Err: Display> Display for TransactionFailure<Err> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            TransactionFailure::FailedAndRolledBackCompletely(err) => {
+                write!(f, "Transaction failed with {err} and rolled back completely.")
+            }
+            TransactionFailure::FailedAndRolledBackPartially {
+                failure,
+                compensation_failure,
+            } => write!(
+                f,
+                "Transaction failed with {failure} and rolled back partially; compensation failed with: {compensation_failure}."
+            ),
+        }
+    }
+}
+
+/// Fallible transaction execution. If any operation fails, all the already executed
+/// operation's compensation actions are executed in reverse order and the transaction
+/// returns with a failure.
+pub fn fallible_transaction<Out, Err: Clone + 'static>(
+    f: impl FnOnce(&mut FallibleTransaction<Err>) -> TransactionResult<Out, Err>,
```

Comment:
```suggestion
    f: impl FnOnce(&mut FallibleTransaction<Err>) -> Result<Out, Err>,
```

I don't the user should be able to generate a `TransactionResult`. I think they can return `Result` and then we will wrap that in `TransactionResult` as appropriate.

Reply:
I did this because the actual transaction result is coming from a call _within_ this function (a `tx.execute`) currently. But I will try to restructure it so it is not necessary (moving the error handling into the outer level)